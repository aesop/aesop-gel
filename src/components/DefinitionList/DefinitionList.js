import React, { Fragment, forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { isObjectPopulatedArray } from '~/utils/objects';
import styles from './DefinitionList.module.css';

const DefinitionList = forwardRef(function DefinitionListRef(
  { className, hasBottomBorder, isVisible, items, theme },
  ref,
) {
  const classSet = cx(
    styles.base,
    styles[theme],
    { [styles.hasBottomBorder]: hasBottomBorder },
    className,
  );
  const termClassSet = cx(styles.term, { [styles.slideIn]: isVisible });
  const descriptionClassSet = cx(styles.description, {
    [styles.slideIn]: isVisible,
  });

  if (!isObjectPopulatedArray(items)) {
    return null;
  }

  return (
    <dl className={classSet} ref={ref}>
      {items
        .filter(({ description, term }) => description && term)
        .map(({ description, id, term }) => (
          <Fragment key={id}>
            <dt className={termClassSet}>{term}</dt>
            <dd className={descriptionClassSet}>{description}</dd>
          </Fragment>
        ))}
    </dl>
  );
});

DefinitionList.propTypes = {
  className: PropTypes.string,
  hasBottomBorder: PropTypes.bool,
  isVisible: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      description: PropTypes.node.isRequired,
      id: PropTypes.string.isRequired,
      term: PropTypes.node.isRequired,
    }),
  ),
  theme: PropTypes.oneOf(['dark', 'light']),
};

DefinitionList.defaultProps = {
  className: undefined,
  hasBottomBorder: false,
  isVisible: true,
  items: [],
  theme: 'dark',
};

export default DefinitionList;
