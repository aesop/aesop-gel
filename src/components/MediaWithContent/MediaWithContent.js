import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Media from './components/Media';
import Content from './components/Content';
import styles from './MediaWithContent.module.css';

const MediaWithContent = forwardRef(
  (
    {
      backgroundColor,
      className,
      content,
      copy,
      foregroundImage,
      hasFullWidthImage,
      isHero,
      isReverse,
      media,
      theme,
    },
    ref,
  ) => {
    const classSet = cx(
      styles.base,
      { [styles.reverse]: isReverse },
      { [styles.hero]: isHero },
      className,
    );

    return (
      <section className={classSet} ref={ref} style={{ backgroundColor }}>
        <Media
          className={styles.media}
          foregroundImage={foregroundImage}
          hasFullWidthImage={hasFullWidthImage}
          isHero={isHero}
          media={media}
        />
        <Content
          content={content}
          copy={copy}
          hasFullWidthImage={hasFullWidthImage}
          isHero={isHero}
          isReverse={isReverse}
          theme={theme}
        />
      </section>
    );
  },
);

MediaWithContent.propTypes = {
  backgroundColor: PropTypes.string,
  className: PropTypes.string,
  content: PropTypes.any,
  copy: PropTypes.shape({
    description: PropTypes.any,
    eyebrow: PropTypes.string,
    heading: PropTypes.string,
    subHeading: PropTypes.string,
  }).isRequired,
  foregroundImage: PropTypes.element,
  hasFullWidthImage: PropTypes.bool,
  isHero: PropTypes.bool,
  isReverse: PropTypes.bool,
  media: PropTypes.element.isRequired,
  theme: PropTypes.oneOf(['dark', 'light']),
};

MediaWithContent.defaultProps = {
  backgroundColor: '#F6F5E8',
  className: undefined,
  content: undefined,
  copy: {
    description: undefined,
    eyebrow: undefined,
    heading: undefined,
    subHeading: undefined,
  },
  foregroundImage: undefined,
  hasFullWidthImage: false,
  isHero: false,
  isReverse: false,
  media: undefined,
  theme: 'dark',
};

export default MediaWithContent;
