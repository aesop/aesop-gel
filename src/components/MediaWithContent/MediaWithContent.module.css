@import 'styles/index';

.base {
  position: relative;

  @media (--viewport-lg) {
    display: flex;
    min-height: 100vh;
    flex-direction: row;
  }

  &.hero {
    @media (--viewport-lg) {
      min-height: 70vh;
    }
  }
}

.reverse {
  @media (--viewport-lg) {
    flex-direction: row-reverse;
  }
}
