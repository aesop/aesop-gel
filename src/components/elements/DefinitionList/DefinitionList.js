import React from 'react';
import cx from 'classnames';
import Props from './DefinitionList.types';
import styles from './DefinitionList.module.css';

const DefinitionList: React.FunctionComponent<Props> = ({
  className,
  items,
}) => {
  const classSet: string = cx(styles.base, className);

  return (
    <dl className={classSet}>
      {items
        .filter(({ description, term }) => description && term)
        .map(({ description, term }, index) => (
          <React.Fragment key={index}>
            <dt className={styles.term}>{term}</dt>
            <dd className={styles.description}>{description}</dd>
          </>
        ))}
    </dl>
  );
};

DefinitionList.defaultProps = {
  items: [],
};

export default DefinitionList;
