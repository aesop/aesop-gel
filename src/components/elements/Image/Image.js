import React from 'react';
import cx from 'classnames';
import get from 'lodash/get';
import { constraintsByViewport } from '~/utils/viewports/viewports';
import Hyperlink from '~/components/elements/Hyperlink';
import { ImageType as Props } from '~/types/';
import styles from './Image.module.css';

const Image: React.FunctionComponent<Props> = ({
  altText,
  className,
  cta,
  large,
  medium,
  small,
  type,
}) => {
  const fullBleedImage = type === 'componentFullBleedImage';
  const isRightAligned = type === 'componentSquareImage';

  const classSet: string = cx(
    styles.base,
    { [styles.fullBleedImage]: fullBleedImage },
    { [styles.isRightAligned]: isRightAligned },
    className,
  );

  const picture = (
    <picture className={cx(styles.picture, className)}>
      {large && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'large.minWidth',
          )}px)`}
          srcSet={large}
        />
      )}

      {medium && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'medium.minWidth',
          )}px)`}
          srcSet={medium}
        />
      )}

      {small && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'small.minWidth',
          )}px)`}
          srcSet={small}
        />
      )}

      <img alt={altText} />
    </picture>
  );

  return cta ? (
    <Hyperlink openInANewWindow={cta.openInANewWindow} url={cta.url}>
      {picture}
    </Hyperlink>
  ) : (
    picture
  );
};

export default Image;
