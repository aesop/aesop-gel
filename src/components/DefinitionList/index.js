/* istanbul ignore file */
import DefinitionList from './DefinitionList';

export default DefinitionList;
