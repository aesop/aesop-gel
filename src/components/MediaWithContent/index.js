/* istanbul ignore file */
import MediaWithContent from './MediaWithContent';

export default MediaWithContent;
