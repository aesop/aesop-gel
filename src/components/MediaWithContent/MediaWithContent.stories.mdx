import { Fragment } from 'react';
import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { boolean, select, text, radios } from '@storybook/addon-knobs';
import DefinitionList from '~/components/DefinitionList';
import Hyperlink from '~/components/Hyperlink';
import Image from '~/components/Image';
import KitList from '~/components/KitList';
import LinkButtonGroup from '~/components/LinkButtonGroup';
import SubNav from '~/components/SubNav';
import Video from '~/components/Video';
import MediaWithContent from './MediaWithContent';
import MediaWithContentFixture from './MediaWithContent.fixture';

<Meta title="Components/MediaWithContent" component={MediaWithContent} />

# MediaWithContent

<Preview>
  <Story name="Base Component">
    {(() => {
      const theme = knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation');
      return (
        <MediaWithContent
          backgroundColor={knobs.text('backgroundColor', null, 'Presentation')}
          content={(() => {
            const options = {
              DefinitionList: {
                Component: DefinitionList,
                props: MediaWithContentFixture.definitionList,
              },
              KitList: {
                Component: KitList,
                props: MediaWithContentFixture.kitList,
              },
              LinkButtonGroup: {
                Component: LinkButtonGroup,
                props: MediaWithContentFixture.linkButtonGroup,
              },
            };
            const componentName = knobs.select('content', Object.keys(options), 'LinkButtonGroup', 'Content');
            const { Component, props } = options[componentName];
            return <Component {...props} theme={theme} />
          })()}
          copy={{
            description: knobs.text(
              'description',
              MediaWithContentFixture.copy.default.description,
              'Copy',
            ),
            eyebrow: knobs.text('eyebrow', MediaWithContentFixture.copy.default.eyebrow, 'Copy'),
            heading: knobs.text('heading', MediaWithContentFixture.copy.default.heading, 'Copy'),
            subHeading: knobs.text('subHeading', MediaWithContentFixture.copy.default.subHeading, 'Copy'),
          }}
          hasFullWidthImage={knobs.boolean('hasFullWidthImage', false, 'Presentation')}
          isHero={knobs.boolean('isHero', false, 'Presentation')}
          isReverse={knobs.boolean('isReverse', false, 'Presentation')}
          media={
            <Image
              altText={MediaWithContentFixture.image.expectations.altText}
              large={MediaWithContentFixture.image.expectations.large}
              medium={MediaWithContentFixture.image.expectations.medium}
              small={MediaWithContentFixture.image.expectations.small}
            />
          }
          theme={theme}
        />
      )
    })()}
  </Story>
</Preview>

## API

<Props of={MediaWithContent} />

## Variations

### With Definition List

<Preview>
  <Story name="Definition List">
    <MediaWithContent
      backgroundColor="#fffeef"
      content={
        <DefinitionList items={MediaWithContentFixture.definitionList.items} />
      }
      copy={{
        eyebrow: MediaWithContentFixture.copy.default.eyebrow,
        heading: MediaWithContentFixture.copy.default.heading,
      }}
      hasFullWidthImage={false}
      isHero={false}
      isReverse={false}
      media={
        <Image
          altText={MediaWithContentFixture.image.default.altText}
          large={MediaWithContentFixture.image.default.large}
          medium={MediaWithContentFixture.image.default.medium}
          small={MediaWithContentFixture.image.default.small}
        />
      }
    />
  </Story>
</Preview>

### With Kit List

<Preview>
  <Story name="With Kit List">
    <MediaWithContent
      content={<KitList items={MediaWithContentFixture.kitList.items} />}
      copy={MediaWithContentFixture.kitList.copy}
      media={
        <Image
          altText={MediaWithContentFixture.image.kit.altText}
          large={MediaWithContentFixture.image.kit.large}
          medium={MediaWithContentFixture.image.kit.medium}
          small={MediaWithContentFixture.image.kit.small}
        />
      }
    />
  </Story>
</Preview>

### With Expectations

<Preview>
  <Story name="With Expectations">
    <MediaWithContent
      copy={MediaWithContentFixture.copy.expectations}
      isReverse={true}
      media={
        <Image
          altText={MediaWithContentFixture.image.expectations.altText}
          large={MediaWithContentFixture.image.expectations.large}
          medium={MediaWithContentFixture.image.expectations.medium}
          small={MediaWithContentFixture.image.expectations.small}
        />
      }
    />
  </Story>
</Preview>

### Hero Image

<Preview>
  <Story name="Hero Image">
    <MediaWithContent
      backgroundColor="#633e43"
      content={(() => {
        const options = {
          SubNav: {
            Component: SubNav,
            props: MediaWithContentFixture.subNav,
          },
          LinkButtonGroup: {
            Component: LinkButtonGroup,
            props: MediaWithContentFixture.linkButtonGroup,
          },
        };
        const componentName = select('content', Object.keys(options), 'LinkButtonGroup');
        const { Component, props } = options[componentName];
        return <Component {...props} theme="light" />
      })()}
      copy={MediaWithContentFixture.copy.expectations}
      hasFullWidthImage={false}
      isHero={true}
      isReverse={true}
      media={
        <Image
          altText={MediaWithContentFixture.image.hero.altText}
          large={MediaWithContentFixture.image.hero.large}
          medium={MediaWithContentFixture.image.hero.medium}
          small={MediaWithContentFixture.image.hero.small}
        />
      }
      theme="light"
    />
  </Story>
</Preview>

### Hero Image Full Width

<Preview>
  <Story name="Hero Image Full Width">
    <MediaWithContent
      backgroundColor="#222e56"
      content={(() => {
        const options = {
          SubNav: {
            Component: SubNav,
            props: MediaWithContentFixture.subNav,
          },
          LinkButtonGroup: {
            Component: LinkButtonGroup,
            props: MediaWithContentFixture.linkButtonGroup,
          },
        };
        const componentName = select('content', Object.keys(options), 'LinkButtonGroup');
        const { Component, props } = options[componentName];
        return <Component {...props} theme="light" />
      })()}
      copy={{
        heading: MediaWithContentFixture.copy.default.heading,
      }}
      hasFullWidthImage={true}
      isHero={true}
      isReverse={true}
      media={
        <Image
          altText={MediaWithContentFixture.image.expectations.altText}
          large={MediaWithContentFixture.image.expectations.large}
          medium={MediaWithContentFixture.image.expectations.medium}
          small={MediaWithContentFixture.image.expectations.small}
        />
      }
      theme="light"
    />
  </Story>
</Preview>

### Hero Image with Foreground Image

<Preview>
  <Story name="Hero Image with Foreground Image">
    <MediaWithContent
      backgroundColor="#f5f3e6"
      content={
        <Hyperlink {...MediaWithContentFixture.hyperlinkButtonDark} />
      }
      copy={MediaWithContentFixture.copy.expectations}
      foregroundImage={
        <Image
          altText={MediaWithContentFixture.image.foreground.altText}
          large={MediaWithContentFixture.image.foreground.large}
          medium={MediaWithContentFixture.image.foreground.medium}
          small={MediaWithContentFixture.image.foreground.small}
        />
      }
      foregroundImageLink={MediaWithContentFixture.foregroundImageLink}
      hasFullWidthImage={false}
      isHero={true}
      isReverse={true}
      media={
        <Image
          altText={MediaWithContentFixture.image.background.altText}
          large={MediaWithContentFixture.image.background.large}
          medium={MediaWithContentFixture.image.background.medium}
          small={MediaWithContentFixture.image.background.small}
        />
      }
      theme="dark"
    />
  </Story>
</Preview>

### Hero Video

<Preview>
  <Story name="Hero Video">
    <MediaWithContent
      backgroundColor="#633e43"
      content={
        <Fragment>
          <LinkButtonGroup
            {...MediaWithContentFixture.linkButtonGroup}
            theme="light"
          />
        </Fragment>
      }
      copy={MediaWithContentFixture.copy.expectations}
      hasFullWidthImage={false}
      isHero={true}
      isReverse={true}
      media={
        <Video
          hasAllowAudio={false}
          hasAutoplay={true}
          hasLoop={true}
          hasPlayInFullScreen={false}
          hasControls={false}
          isFullWidth={false}
          large={MediaWithContentFixture.video.source}
          medium={MediaWithContentFixture.video.source}
          small={MediaWithContentFixture.video.source}
        />
      }
      theme="light"
    />
  </Story>
</Preview>

### Hero Video full width

<Preview>
  <Story name="Hero Video full width">
    <MediaWithContent
      backgroundColor="#000"
      copy={MediaWithContentFixture.copy.default}
      hasFullWidthImage={true}
      isHero={true}
      isReverse={true}
      media={
        <Video
          hasAllowAudio={false}
          hasAutoplay={true}
          hasLoop={true}
          hasPlayInFullScreen={false}
          hasControls={false}
          isFullWidth={false}
          large={MediaWithContentFixture.video.source}
          medium={MediaWithContentFixture.video.source}
          small={MediaWithContentFixture.video.source}
        />
      }
      theme="light"
    />
  </Story>
</Preview>
