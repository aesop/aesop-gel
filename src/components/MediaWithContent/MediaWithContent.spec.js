import React from 'react';
import renderer from 'react-test-renderer';
import DefinitionList from '~/components/DefinitionList';
import Image from '~/components/Image';
import MediaWithContent from './MediaWithContent';
import MediaWithContentFixture from './MediaWithContent.fixture';

describe('<MediaWithContent />', () => {
  it('should be defined', () => {
    expect(MediaWithContent).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <MediaWithContent
          content={
            <DefinitionList
              items={MediaWithContentFixture.definitionList.items}
            />
          }
          copy={MediaWithContentFixture.copy.expectations}
          media={<Image {...MediaWithContentFixture.image.default} />}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
