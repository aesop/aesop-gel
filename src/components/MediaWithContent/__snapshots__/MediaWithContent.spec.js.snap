// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaWithContent /> renders base component correctly 1`] = `
<div
  className="base"
  style={
    Object {
      "backgroundColor": "#F6F5E8",
    }
  }
>
  <div
    className="media base undefined"
  >
    <figure
      className="base dark figure"
    >
      <picture
        className="base picture dark"
      >
        <source
          media="(min-width: 1025px)"
          srcSet="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
        />
        <source
          media="(min-width: 640px)"
          srcSet="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
        />
        <source
          media="(min-width: 0px)"
          srcSet="https://www.aesop.com/medias/Aesop-Skin-Classic-Purifying-Facial-Exfoliant-Paste-Texture-50-50-Mobile-640x640px.jpg?context=bWFzdGVyfGltYWdlc3wyNDg1OHxpbWFnZS9qcGVnfGltYWdlcy9oMjIvaDBkLzkxMTYyMjg1MTc5MTguanBnfDJjZDUzMjI4N2EzZDgyYzE1ZTk3MTM4OWFjODU5N2Q2NjQ4YTAxNzcwZmE4YjBlZDQ2YTJmYTRhOTI5ZWQyY2I"
        />
        <img
          alt="alt image"
        />
      </picture>
    </figure>
  </div>
  <div
    className="base notHero hasHalfWidthImage dark"
  >
    <div
      className="innerWrapper"
    >
      <header
        className="header base"
      >
        <h2
          className="base flush serifFont xLarge dark heading"
        >
          Fusce et dui nisi
        </h2>
        <h3
          className="base flush mediumWeightFont xXSmall dark eyebrow eyebrow"
        >
          Quisque blandit est erat
        </h3>
      </header>
      <div
        className="content"
      >
        <div
          className="set dark description"
        >
          Aliquam lectus purus, gravida id metus dignissim, efficitur tempus lacus. Maecenas a cursus diam. Vivamus ultrices nisi non ex placerat, vel elementum sapien condimentum.
        </div>
        <dl
          className="base dark"
        >
          <dt
            className="term slideIn"
          >
            Aliquam
          </dt>
          <dd
            className="description slideIn"
          >
            Proin vulputate eros quis lectus pretium gravida.
          </dd>
          <dt
            className="term slideIn"
          >
            Etiam eu augue quam
          </dt>
          <dd
            className="description slideIn"
          >
            Pulvinar nulla enim, eu elementum lacus vehicula sed.
          </dd>
          <dt
            className="term slideIn"
          >
            Duis at elit
          </dt>
          <dd
            className="description slideIn"
          >
            Sed ac ipsum vel risus aliquam rutrum.
          </dd>
        </dl>
      </div>
    </div>
  </div>
</div>
`;
